File: 3_Stack/dailyTemperatures.js
// Function to calculate the number of days until warmer temperatures for each day
var dailyTemperatures = function(temperatures, hottest = 0) {
    // Initialize an array to store the result with default values of 0
    const days = new Array(temperatures.length).fill(0);

    // Iterate through each day in reverse order
    for (let day = (temperatures.length - 1); (0 <= day); day--) {
        // Get the temperature for the current day
        const temperature = temperatures[day];

        // Check if the current temperature is hotter than or equal to the hottest temperature
        const isHotter = hottest <= temperature;
        if (isHotter) {
            hottest = temperature;
            continue; // Move to the next iteration if the current temperature is hotter or equal
        }

        // If the current temperature is not hotter, perform a search for the next warmer day
        search(temperatures, day, temperature, days);
    }

    // Return the array representing the number of days until warmer temperatures for each day
    return days;
}

// Function to search for the next warmer day
const search = (temperatures, day, temperature, days, dayCount = 1) => {
    // Helper function to check if the temperature on the next day is hotter than the current temperature
    const isHotter = () => temperatures[day + dayCount] <= temperature;

    // Continue searching until a warmer day is found
    while (isHotter()) {
        // Increment the day count by the number of days until warmer temperatures from the next day
        dayCount += days[day + dayCount];
    }

    // Update the result array with the number of days until warmer temperatures for the current day
    days[day] = dayCount;
}


// implementation
console.log("Leetcode 739: Daily Temperatures\n");

// Example 1
const example1 = [73, 74, 75, 71, 69, 72, 76, 73];
console.log(`Example 1: temperatures = [${example1}]`);
console.log("Output:", dailyTemperatures(example1));

// Example 2
const example2 = [30, 40, 50, 60];
console.log(`Example 2: temperatures = [${example2}]`);
console.log("Output:", dailyTemperatures(example2));

// Example 3
const example3 = [30, 60, 90];
console.log(`Example 3: temperatures = [${example3}]`);
console.log("Output:", dailyTemperatures(example3));

File: 03_Stack/daily_temperatures.py
from typing import List

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # Initialize a list to store the result with default values of 0
        res = [0] * len(temperatures)

        # Initialize a stack to keep track of temperatures and their indices
        stack = []  # Stack elements: [temperature, index]

        # Iterate through each temperature and its index
        for i, t in enumerate(temperatures):
            # Check if the current temperature is greater than the temperature at the top of the stack
            while stack and t > stack[-1][0]:
                # If it is greater, pop the stack and update the result for the popped index
                stackTemp, stackInd = stack.pop()
                res[stackInd] = i - stackInd

            # Push the current temperature and its index onto the stack
            stack.append((t, i))

        # Return the list of results representing the number of days until warmer temperatures
        return res
    
# implementation
answer = Solution()

print("Leetcode 739: Daily Temperatures\n")

# Example 1
example_1 = [73, 74, 75, 71, 69, 72, 76, 73]
print(f"Example 1: temperatures = {example_1}")
print("Output:", answer.dailyTemperatures(example_1))

# Example 2
example_2 = [30, 40, 50, 60]
print(f"Example 2: temperatures = {example_2}")
print("Output:", answer.dailyTemperatures(example_2))

# Example 3
example_3 = [30, 60, 90]
print(f"Example 3: temperatures = {example_3}")
print("Output:", answer.dailyTemperatures(example_3))